File: java/src/test/java/Edition055_Device_Log_Streaming.java
import io.appium.java_client.AppiumDriver;
import org.java_websocket.client.WebSocketClient;
import org.java_websocket.handshake.ServerHandshake;
import org.junit.Test;
import org.openqa.selenium.remote.DesiredCapabilities;
import java.net.MalformedURLException;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.URL;



public class Edition055_Device_Log_Streaming {

    private String ANDROID_APP = "https://github.com/cloudgrey-io/the-app/releases/download/v1.8.1/TheApp-v1.8.1.apk";
    private String IOS_APP = "https://github.com/cloudgrey-io/the-app/releases/download/v1.6.1/TheApp-v1.6.1.app.zip";

    private AppiumDriver driver;

    public class LogClient extends WebSocketClient {

        public LogClient( URI serverURI ) {
            super(serverURI);
        }

        @Override
        public void onOpen(ServerHandshake handshakedata) {
            System.out.println("WEBSOCKET OPENED");
        }

        @Override
        public void onMessage(String message) {
            System.out.println(message);
        }

        @Override
        public void onClose(int code, String reason, boolean remote) {
            System.out.println("Connection closed, log streaming has stopped");
        }

        @Override
        public void onError(Exception ex) {
            ex.printStackTrace();
            // if the error is fatal then onClose will be called additionally
        }
    }

    @Test
    public void streamAndroidLogs() throws URISyntaxException, MalformedURLException {
        DesiredCapabilities caps = new DesiredCapabilities();
        caps.setCapability("platformName", "Android");
        caps.setCapability("deviceName", "Android Emulator");
        caps.setCapability("automationName", "UiAutomator2");
        caps.setCapability("app", ANDROID_APP);

        driver = new AppiumDriver(new URL("http://localhost:4723/wd/hub"), caps);

        LogClient logClient = new LogClient(new URI( "ws://localhost:4723/ws/session/" + driver.getSessionId() + "/appium/device/logcat"));

        driver.executeScript("mobile:startLogsBroadcast");

        logClient.connect();

        try { Thread.sleep(5000); } catch (Exception ign) {} // logs printed to stdout while we're sleeping.

        driver.executeScript("mobile:stopLogsBroadcast");
        driver.quit();
    }

    @Test
    public void streamIOSLogs() throws URISyntaxException, MalformedURLException {
        DesiredCapabilities caps = new DesiredCapabilities();
        caps.setCapability("platformName", "iOS");
        caps.setCapability("platformVersion", "12.1");
        caps.setCapability("deviceName", "iPhone XS");
        caps.setCapability("automationName", "XCUITest");
        caps.setCapability("app", IOS_APP);

        driver = new AppiumDriver(new URL("http://localhost:4723/wd/hub"), caps);

        LogClient logClient = new LogClient( new URI("ws://localhost:4723/ws/session/" + driver.getSessionId() + "/appium/device/syslog"));

        driver.executeScript("mobile:startLogsBroadcast");

        logClient.connect();

        try { Thread.sleep(5000); } catch (Exception ign) {} // logs printed to stdout while we're sleeping.

        driver.executeScript("mobile:stopLogsBroadcast");
        driver.quit();
    }
}

File: javascript/Edition055_Device_Log_Streaming.js
let test = require('ava')
let { remote } = require('webdriverio')
let B = require('bluebird')
let websocket = require('websocket-stream')

let driver

test('stream Android logcat logs', async t => {
  driver = await remote({
    hostname: 'localhost',
    port: 4723,
    path: '/wd/hub',
    capabilities: {
      platformName: 'Android',
      deviceName: 'Android Emulator',
      automationName: 'UiAutomator2',
      app: 'https://github.com/cloudgrey-io/the-app/releases/download/v1.8.1/TheApp-v1.8.1.apk'
    },
    logLevel: 'error'
  })

  let logStream = websocket(`ws://localhost:4723/ws/session/${driver.sessionId}/appium/device/logcat`)
  logStream.pipe(process.stdout)

  logStream.on('finish', () => {
    console.log('Connection closed, log streaming has stopped')
  })

  driver.executeScript('mobile:startLogsBroadcast', [])

  await B.delay(5000)

  driver.executeScript('mobile:stopLogsBroadcast', [])

  await driver.deleteSession()
  t.pass()
})

test('stream iOS system logs', async t => {
  driver = await remote({
    hostname: 'localhost',
    port: 4723,
    path: '/wd/hub',
    capabilities: {
      platformName: 'iOS',
      platformVersion: '12.1',
      deviceName: 'iPhone XS',
      automationName: 'XCUITest',
      app: 'https://github.com/cloudgrey-io/the-app/releases/download/v1.6.1/TheApp-v1.6.1.app.zip'
    },
    logLevel: 'error'
  })

  let logStream = websocket(`ws://localhost:4723/ws/session/${driver.sessionId}/appium/device/syslog`)
  logStream.pipe(process.stdout)

  logStream.on('finish', () => {
    console.log('Connection closed, log streaming has stopped')
  })

  driver.executeScript('mobile:startLogsBroadcast', [])

  await B.delay(5000)

  driver.executeScript('mobile:stopLogsBroadcast', [])

  await driver.deleteSession()
  t.pass()
})
